### PR TITLE
Adding loglevel for overflow strategy #25949

### PR DESCRIPTION
**Pull request by nvollmar**
Adds optional log level to overflow strategy to allow override of the default log level for dropping items or failing the stream when using `Source.actorRef`

### CODE REVIEW

**Comment by nvollmar** [comment]
@patriknw I've run into issues with your code validator:
`[error]`  * the type hierarchy of object akka.stream.OverflowStrategies#Fail is different in current version. Missing types {akka.stream.DelayOverflowStrategy,akka.stream.OverflowStrategy}
[error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("akka.stream.OverflowStrategies$Fail$")
[error]  * method productElement(Int)java.lang.Object in object akka.stream.OverflowStrategies#Fail does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.OverflowStrategies#Fail.productElement")
[error]  * method productArity()Int in object akka.stream.OverflowStrategies#Fail does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.OverflowStrategies#Fail.productArity")
[error]  * method canEqual(java.lang.Object)Boolean in object akka.stream.OverflowStrategies#Fail does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.OverflowStrategies#Fail.canEqual")
[error]  * method productIterator()scala.collection.Iterator in object akka.stream.OverflowStrategies#Fail does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.OverflowStrategies#Fail.productIterator")
[error]  * method productPrefix()java.lang.String in object akka.stream.OverflowStrategies#Fail does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.stream.OverflowStrategies#Fail.productPrefix")
[error]  * method toString()java.lang.String in object akka.stream.OverflowStrategies#Fail is declared final in current version
[error]    filter with: ProblemFilters.exclude[FinalMethodProblem]("akka.stream.OverflowStrategies#Fail.toString")`
Since those classes are private to akka this change should not break any api, do you have a hint?

File: akka-stream/src/main/scala/akka/stream/OverflowStrategy.scala [hunk 1]
```diff
@@ -59,34 +61,67 @@ object OverflowStrategy {
...
    * the new element.
    */
-  def dropHead: OverflowStrategy = DropHead
+  def dropHead: OverflowStrategy = DropHead(Logging.DebugLevel)
+
+  /**
+   * If the buffer is full when a new element arrives, drops the oldest element from the buffer to make space for
+   * the new element.
+   */
+  def dropHead(logLevel: LogLevel): OverflowStrategy = DropHead(logLevel)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/OverflowStrategy.scala]: an alternative API could be
```
OverflowStrategy.dropHead.withLogLevel(Attributes.LogLevels.Warning)
```

File: akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala [hunk 2]
```diff
@@ -62,23 +62,23 @@ import scala.util.control.NonFatal
...
-          case DropTail ⇒
+          case _: DropTail ⇒
             buffer.dropTail()
             enqueueAndSuccess(offer)
-          case DropBuffer ⇒
+          case _: DropBuffer ⇒
             buffer.clear()
             enqueueAndSuccess(offer)
-          case DropNew ⇒
+          case _: DropNew ⇒
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala]: then we should log here also, using the provided log level

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala]: and in all other places where OverflowStrategy is used

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 3]
```diff
@@ -1690,7 +1690,7 @@ private[stream] object Collect {
     }
 
     def pullCondition: Boolean =
-      strategy != Backpressure || buffer.used < size
+      !strategy.isInstanceOf[Backpressure] || buffer.used < size
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: perhaps add `private[akka] def isBackpressure: Boolean` to `OverflowStrategy`

File: akka-stream/src/main/scala/akka/stream/OverflowStrategy.scala (hunk 1 above)
Comment (nvollmar) [reply akka-stream/src/main/scala/akka/stream/OverflowStrategy.scala]: Yes, that will result in a simpler/less redundant API

**Comment by patriknw** [comment]
> Since those classes are private to akka this change should not break any api, do you have a hint?

That is right. It's ok to add those filters to akka-stream/src/main/mima-filters/2.5.18.backwards.excludes

**Comment by nvollmar** [comment]
@patriknw I've addressed all your comments and updated the pull request

File: akka-stream/src/main/scala/akka/stream/OverflowStrategy.scala [hunk 4]
```diff
@@ -6,13 +6,17 @@ package akka.stream
...
+import akka.event.Logging.LogLevel
 
 /**
  * Represents a strategy that decides how to deal with a buffer of time based operator
  * that is full but is about to receive a new element.
  */
 @DoNotInherit
-sealed abstract class DelayOverflowStrategy extends Serializable
+sealed abstract class DelayOverflowStrategy extends Serializable {
+  private[akka] def isBackpressure: Boolean
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/OverflowStrategy.scala]: a few more markers:
```
    /** INTERNAL API */
    @InternalApi private[akka]
```

File: akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala [hunk 5]
```diff
@@ -40,7 +40,7 @@ import scala.util.control.NonFatal
   override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
     val completion = Promise[Done]
 
-    val stageLogic = new GraphStageLogic(shape) with OutHandler with SourceQueueWithComplete[T] {
+    val stageLogic = new GraphStageLogic(shape) with OutHandler with SourceQueueWithComplete[T] with StageLogging {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/QueueSource.scala]: better logger name if you override `logSource` in `StageLogging`
```
override protected def logSource = classOf[QueueSource]
```

in all places

**Review by patriknw — COMMENTED** [review]
looking good, just some additional minor things

**Review by patriknw — APPROVED** [review]
LGTM, thanks for quick follow up

**Comment by patriknw** [comment]
OK TO TEST

**Review by chbatey — COMMENTED** [review]
Need to move the mima excludes then LGTM

**Review by chbatey — APPROVED** [review]
LGTM

**Comment by chbatey** [comment]
Thanks @nvollmar

**Pull request merged**